### DISCUSSION TITLE
在高并发场景下，有管理连接数的技术方案吗？大家是用连接池控制连接数还是有其他方式呢？

### DISCUSSION THREAD

**Discussion by zyy20191** [milvus-io/milvus#18011] [Q&A and General discussion]
我是用python对Mlivus服务进行操作，想问问在10w个用户同时访问服务的情况下，由于最大连接数为65536，那么需要一个连接进行多个查询，针对这个场景我有如下几个问题：
1、Mlivus有连接池的概念吗，有的话怎么用呢？
2、如果没有连接池的概念，那怎么管理连接数呢，官方有好的方案吗？大家是怎么管理的呢？
3、一个连接进行多个查询，是串行查询吗，后面想查询的用户需要等待？

**Comment by sunmingqiaa** [comment]
这么好的问题怎么没有回复呢

**Reply by xiaofan-luan** [reply]
@czs007

**Reply by zyy20191** [reply]
你可以看看问题#18034和#22264，在另外的帖子下有作相关回答

**Comment by yangyu6** [comment]
有结论吗老哥？我看其他帖子也没看出来结论，到底需不需要池化 milvusClient？

**Reply by yhmo** [reply]
pymilvus目前没有提供连接池，用户可以在业务端自行实现。

milvusClient的请求发送路径有两条，如下图所示：
一种是dml请求，包括delete/insert/upsert，途经client ---> proxy ---> pulsar/kafka --->datanode/querynode
第二种是查询请求，包括search/query/hybrid_search，途经client--->proxy--->querynode
![dml_tasks](https://github.com/user-attachments/assets/610cdea4-5417-4af9-9320-d666c08da85b)

相应的，对于dml请求，每个proxy中都有一个dml tasks queue顺序处理接收到的dml请求。对于查询请求，每个proxy中都有一个query tasks queue顺序处理接收到的查询请求。
所以，就算业务端有10000个milvusclient同时发各种请求，proxy也是按照接收到的先后顺序处理。dml请求处理的快慢由后端的pulsar/kafka决定。查询请求处理的快慢由后端的querynode决定。

**Reply by yangyu6** [reply]
多谢，我使用的是 java-sdk 2.5.3 版本，有区别吗?
  ```
  <dependency>
            <groupId>io.milvus</groupId>
            <artifactId>milvus-sdk-java</artifactId>
  </dependency>
```

**Reply by yhmo** [reply]
Java SDK提供了连接池，可以帮助你建立一个可自动调节的池子。池子里会预先创建一些MilvusClient对象，当你的请求比较密集时，会自动增加MilvusClient对象，当你的请求稀疏时，会自动释放一部分空闲的MilvusClient对象。
这里有使用示例：
https://github.com/milvus-io/milvus-sdk-java/blob/v2.5.3/examples/main/java/io/milvus/v2/ClientPoolExample.java

**Reply by yangyu6** [reply]
非常感谢您，解决了我的问题。
我在 java-sdk的仓库中提了问题，没有人响应。借楼请教一下您。
milvus版本是 2.3.1，删除时报错，按照filter条件search可以搜到记录
https://github.com/milvus-io/milvus-sdk-java/issues/1314

**Reply by yhmo** [reply]
delete接口支持表达式是在milvus 2.3.2才有，在2.3.2之前的版本，只能根据主键delete
